pkg/logger/json_test.go

```diff
@@ -0,0 +1,254 @@
+/*
+Copyright 2022 CodeNotary, Inc. All rights reserved.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+	http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package logger
+
+import (
+	"bytes"
+	"encoding/json"
+	"errors"
+	"fmt"
+	"io/ioutil"
+	"os"
+	"runtime"
+	"testing"
+	"time"
+
+	"github.com/stretchr/testify/assert"
+	"github.com/stretchr/testify/require"
+)
+
+func TestJSONLogger(t *testing.T) {
+	t.Run("json log", func(t *testing.T) {
+		var buf bytes.Buffer
+		logger := NewJsonLogger(&LoggerOptions{
+			Name:   "test",
+			Output: &buf,
+		})
+
+		logger.Infof("test call", "user", "foo")
+
+		b := buf.Bytes()
+
+		var raw map[string]interface{}
+		if err := json.Unmarshal(b, &raw); err != nil {
+			t.Fatal(err)
+		}
+
+		assert.Equal(t, "test call", raw["message"])
+		assert.Equal(t, "foo", raw["user"])
```

Comment:
I agree with you. Since most of the golang loggers are built to log a message, and then KV pairs, I was confused whether I should add it as Info rather than Infof. I am thinking I'll keep the Infof to log a formatted message, and change the existing function to a Info, best of both worlds

Reply:
At this point I believe we're only interested in a set of predefined parameters for the log (like module name, call point, log level).
Currently all arguments passed to the `logger.Infof(...)` are only about formatting the message thus I think we will have to leave it that way to not break existing code.

I was also thinking about marking some of those arguments as named fields, e.g.:

```go
logger.Infof("test call from %s to method %s", logger.named("from", "user"), "foo")
```

or some extra builder-like approach (we're using this in TrustCenter backend):

```go
logger.Arg("name", "value").Arg("name2", 46).Infof("test call from %s to method %s", "user", "foo")
```

but that's a much a much larger code change.

Maybe let's just start with a backwards-compatible change without user-specified extra arguments?

Reply:
I agree with you @byo , I have made the Infof backward compatible, but also kept Info to log args, similar to how [zap](https://github.com/uber-go/zap) does

```
sugar.Infow("failed to fetch URL",
  // Structured context as loosely typed key-value pairs.
  "url", url,
  "attempt", 3,
  "backoff", time.Second,
)
sugar.Infof("Failed to fetch URL: %s", url)
```